### PR TITLE
Create SignalR & BackgroundService article

### LINKED ISSUE #8476: SignalR: Add BackgroundService example with IHubContext doc
It's pretty common to write to invoke SignalR clients from a background service. We should update the HubContext document to show the recommended pattern for this in ASP.NET Core applications.

E.g.

https://github.com/davidfowl/UT3/blob/fb12e182d42d2a5a902c1979ea0e91b66fe60607/UTT/Scavenger.cs#L9-L40
 
cc: @anurse  @mikaelm12

### PR DESCRIPTION
**Pull request by bradygaster**
fixes #8476

### CODE REVIEW

File: aspnetcore/signalr/background-services.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ms.author: bradyg
+ms.custom: mvc
+ms.date: 01/31/2018
+uid: signalr/background-services
+---
+# Hosting SignalR in Background Services
+
+By [Brady Gaster](https://twitter.com/bradygaster)
+
+This article provides guidance for hosting SignalR Hubs using background worker process hosted with ASP.NET Core, and how to send messages to connected clients from within a .NET Core [BackgroundService](/dotnet/api/microsoft.extensions.hosting.backgroundservice).
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
This article provides guidance for:

* Hosting SignalR Hubs using a background worker process hosted with ASP.NET Core.
* Sending messages to connected clients from within a .NET Core [BackgroundService](xref:Microsoft.Extensions.Hosting.BackgroundService).
```

File: aspnetcore/signalr/background-services.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/background-services/sample/) [(how to download)](xref:index#how-to-download-a-sample)
+
+## Wiring up SignalR during Startup
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
## Wire up SignalR during startup
```

File: aspnetcore/signalr/background-services.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `ConfigureServices` method of a `Startup` class, calling `services.AddSignalR` adds the required services to the ASP.NET Core DI layer to support SignalR. Then, in `Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
```

File: aspnetcore/signalr/background-services.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+[!code-csharp[Startup](background-service/sample/Server/Startup.cs?name=Startup)]
+
+In the example code, the `ClockHub` class implements the `Hub<T>` class to create a strongly-typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
In the preceding example, the `ClockHub` class implements the `Hub<T>` class to create a strongly typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.
```

File: aspnetcore/signalr/background-services.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+> For more information on strongly-typed Hubs, see the article [Use hubs in SignalR for ASP.NET Core](/aspnet/core/signalr/hubs?view=aspnetcore-2.2#strongly-typed-hubs).
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]: Make this sentence part of the previous paragraph.

Prefer xref-style links, since those are validated during the build process. The value following `xref:` is the `uid` value in the targeted doc's metadata.

```suggestion
> For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).
```

File: aspnetcore/signalr/background-services.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+[!code-csharp[Startup](background-service/sample/Server/ClockHub.cs?name=ClockHub)]
+
+The interface used by the strongly-typed `ClockHub` is the `IClock` interface.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
The interface used by the strongly typed `ClockHub` is the `IClock` interface.
```

File: aspnetcore/signalr/background-services.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+[!code-csharp[Startup](background-service/sample/HubServiceInterfaces/IClock.cs?name=IClock)]
+
+## Calling a SignalR Hub from a background service
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
## Call a SignalR Hub from a background service
```

File: aspnetcore/signalr/background-services.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+During startup, the `Worker` class, a `BackgroundService`, is wired up using `AddHostedService`.
+
+```csharp
+services.AddHostedService<Worker>();
+```
+
+Since SignalR is also wired up during the `Startup` phase in which each Hub is attached to an individual endpoint in ASP.NET Core's HTTP request pipeline, each Hub is represented by an `IHubContext<T>` on the server. Using ASP.NET Core's dependency injection features, other classes instantiated by the hosting layer - like `BackgroundService` classes or MVC Controller classes or even Razor page models - can get references to server-side Hubs by accepting instances of  `IHubContext<ClockHub, IClock>` during construction.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
Since SignalR is also wired up during the `Startup` phase, in which each Hub is attached to an individual endpoint in ASP.NET Core's HTTP request pipeline, each Hub is represented by an `IHubContext<T>` on the server. Using ASP.NET Core's DI features, other classes instantiated by the hosting layer, like `BackgroundService` classes, MVC Controller classes, or Razor page models, can get references to server-side Hubs by accepting instances of `IHubContext<ClockHub, IClock>` during construction.
```

File: aspnetcore/signalr/background-services.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+[!code-csharp[Startup](background-service/sample/Server/Worker.cs?name=Worker)]
+
+As the `ExecuteAsync` method is called iteratively in the background service, the server's current date & time are sent to the connected clients using the `ClockHub`.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
As the `ExecuteAsync` method is called iteratively in the background service, the server's current date and time are sent to the connected clients using the `ClockHub`.
```

File: aspnetcore/signalr/background-services.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+## Reacting to SignalR events with Background Services
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
## React to SignalR events with background services
```

File: aspnetcore/signalr/background-services.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+Using the [.NET Client for SignalR](/aspnet/core/signalr/dotnet-client), a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client applications do using the JavaScript and .NET Clients for SignalR.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
Using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR.
```

File: aspnetcore/signalr/background-services.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+During the `IHostedService.StopAsync` method, the `HubConnection` is disposed of asynchronously.
+
+[!code-csharp[StopAsync method](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=StopAsync)]
+
+## Related resources
+
+* [Get started](xref:tutorials/signalr)
+* [Hubs](xref:signalr/hubs)
+* [Publish to Azure](xref:signalr/publish-to-azure-web-app)
+* [Strongly-typed Hubs](/aspnet/core/signalr/hubs?view=aspnetcore-2.2#strongly-typed-hubs)
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
* [Strongly typed Hubs](xref:signalr/hubs#strongly-typed-hubs)
```

File: aspnetcore/signalr/background-services.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+In the `IHostedService.StartAsync` implementation, the `HubConnection` is started asynchronously.
+
+[!code-csharp[StartAsync method](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=StartAsync)]
+
+During the `IHostedService.StopAsync` method, the `HubConnection` is disposed of asynchronously.
+
+[!code-csharp[StopAsync method](background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs?name=StopAsync)]
+
+## Related resources
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
## Additional resources
```

File: aspnetcore/signalr/background-services.md (hunk 4 above)
Comment (bradygaster) [reply aspnetcore/signalr/background-services.md]: i tried to avoid terms like "preceding" and "below" per feedback i've consistently received from @tdykstra and switched to using "example code" in most cases. Is there a convention for referring to the example code?

File: aspnetcore/signalr/background-services.md (hunk 5 above)
Comment (bradygaster) [reply aspnetcore/signalr/background-services.md]: ah, the xref. i will refer to this in the future as a reference on how to use... uh... *references*. 👍

File: aspnetcore/signalr/background-services.md (hunk 10 above)
Comment (bradygaster) [reply aspnetcore/signalr/background-services.md]: If i had referred to the class name, would you have asked me to leave it as BackgroundServices? Or, had i done that, would the recommendation had been to split it up and make it lower-case? Just curious.

File: aspnetcore/signalr/background-services.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+## Wire up SignalR during startup
+
+Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
+
+[!code-csharp[Startup](background-service/sample/Server/Startup.cs?name=Startup)]
+
+In the preceding example, the `ClockHub` class implements the `Hub<T>` class to create a strongly typed Hub. The `ClockHub` has been configured in the `Startup` class to respond to requests at the endpoint `/hubs/clock`.
+
+> For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]:
```suggestion
For more information on strongly typed Hubs, see [Use hubs in SignalR for ASP.NET Core](xref:signalr/hubs#strongly-typed-hubs).
```

File: aspnetcore/toc.yml [hunk 15]
```diff
@@ -311,6 +311,8 @@
           href: /azure/azure-signalr/signalr-overview
         - name: Redis backplane
           uid: signalr/redis-backplane
+        - name: SignalR with Background Services
```
Comment (scottaddie) [inline aspnetcore/toc.yml]:
```suggestion
        - name: SignalR with background services
```

File: aspnetcore/signalr/background-service/sample/Clients.ConsoleOne/Program.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+            var connection = new HubConnectionBuilder()
+                .WithUrl(Strings.HubUrl)
+                .Build();
+
+            connection.On<DateTime>(Strings.Events.TimeSent, (dateTime) =>
+            {
+                Console.WriteLine(dateTime.ToString());
+            });
+
+            while (true)
```
Comment (davidfowl) [inline aspnetcore/signalr/background-service/sample/Clients.ConsoleOne/Program.cs]: Maybe add a comment here describing the loop here is just to make sure the server is running.

File: aspnetcore/signalr/background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+            _logger.LogInformation($"{currentTime.ToShortTimeString()}");
+
+            return Task.CompletedTask;
+        }
+#endregion
+
+#region StartAsync
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            while (true)
```
Comment (davidfowl) [inline aspnetcore/signalr/background-service/sample/Clients.ConsoleTwo/ClockHubClient.cs]: Maybe add a comment here describing the loop here is just to make sure the server is running.

File: aspnetcore/signalr/background-service/sample/Server/ClockHub.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+using Microsoft.AspNetCore.SignalR;
+
+namespace Server
+{
+#region ClockHub
+    public class ClockHub : Hub<IClock>
+    {
+        public void SendTimeToClients(DateTime dateTime)
+        {
+            Clients.All.ShowTime(dateTime);
```
Comment (davidfowl) [inline aspnetcore/signalr/background-service/sample/Server/ClockHub.cs]: This should be a Task returning method. Make it async and await this call.

**Comment by bradygaster** [comment]
fixes #8476

File: aspnetcore/signalr/background-service/sample/Server/ClockHub.cs (hunk 18 above)
Comment (bradygaster) [reply aspnetcore/signalr/background-service/sample/Server/ClockHub.cs]: Fixed. Realized the Program.cs file had also been obliterated on some commit, so adding it back with these changes.

File: aspnetcore/signalr/background-services.md [hunk 19]
```diff
@@ -8,47 +8,50 @@ ms.custom: mvc
...
 
 [View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/background-services/sample/) [(how to download)](xref:index#how-to-download-a-sample)
 
-## Wiring up SignalR during Startup
+## Wire up SignalR during startup
 
-Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `ConfigureServices` method of a `Startup` class, calling `services.AddSignalR` adds the required services to the ASP.NET Core DI layer to support SignalR. Then, in `Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
+Hosting ASP.NET Core SignalR Hubs in the context of a background worker process is identical to that of hosting a Hub in an ASP.NET Core web app. In the `Startup.ConfigureServices` method, calling `services.AddSignalR` adds the required services to the ASP.NET Core Dependency Injection (DI) layer to support SignalR. In `Startup.Configure`, the `UseSignalR` method is called to wire up the Hub endpoint(s) in the ASP.NET Core request pipeline.
```
Comment (tdykstra) [inline aspnetcore/signalr/background-services.md]: ~~that of~~

File: aspnetcore/signalr/background-services.md [hunk 20]
```diff
@@ -8,47 +8,50 @@ ms.custom: mvc
...
 [!code-csharp[Startup](background-service/sample/Server/Worker.cs?name=Worker)]
 
-As the `ExecuteAsync` method is called iteratively in the background service, the server's current date & time are sent to the connected clients using the `ClockHub`.
+As the `ExecuteAsync` method is called iteratively in the background service, the server's current date and time are sent to the connected clients using the `ClockHub`.
 
-## Reacting to SignalR events with Background Services
+## React to SignalR events with background services
 
-Using the [.NET Client for SignalR](/aspnet/core/signalr/dotnet-client), a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client applications do using the JavaScript and .NET Clients for SignalR.
+Using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR.
```
Comment (tdykstra) [inline aspnetcore/signalr/background-services.md]: This sentence is hard to follow:
* Should "implementor" be "implementation"?
* Using the .NET client ... much like ... .NET client apps do ... using .NET clients [sounds circular]
* Could "can also be used to connect" be changed to "can connect"?
* "like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR"
  * confusing combination of "or" and "and"
  * does this mean to say "like an HTML page can do by using the JavaScript client, and like a .NET desktop app can do by using the .NET client"?

File: aspnetcore/signalr/background-services.md [hunk 21]
```diff
@@ -51,7 +53,7 @@ As the `ExecuteAsync` method is called iteratively in the background service, th
 
 ## React to SignalR events with background services
 
-Using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementor can also be used to connect to SignalR Hubs and respond to events, much like HTML or .NET client apps do using the JavaScript and .NET clients for SignalR.
+Like an HTML using the JavaScript client for SignalR or a .NET desktop app can do using the using the <xref:signalr/dotnet-client>, a `BackgroundService` or `IHostedService` implementation can also be used to connect to SignalR Hubs and respond to events.
```
Comment (scottaddie) [inline aspnetcore/signalr/background-services.md]: I suggest changing this first sentence to begin with "Like a Single Page App using..." instead of "Like an HTML using...".

File: aspnetcore/signalr/background-services.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
+---
+title: Host ASP.NET Core SignalR in background services
+author: bradygaster
+description: Learn how to send messages to SignalR clients from .NET Core BackgroundService classes.
+monikerRange: '>= aspnetcore-2.2'
+ms.author: bradyg
+ms.custom: mvc
+ms.date: 01/31/2018
```
Comment (BrennanConroy) [inline aspnetcore/signalr/background-services.md]: 2019
